sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/BatchingParDoTest.java

```diff
@@ -0,0 +1,284 @@
+  private static BatchingParDo<String, String, String> batchingParDo;
+  private static Instant startInstant;
+  @Rule public transient TestPipeline pipeline = TestPipeline.create();
+
+  @BeforeClass
+  public static void initialize() {
+    startInstant = new Instant(0);
+    data = createTestData();
+    perBatchFn =
+        new SimpleFunction<Iterable<String>, Iterable<String>>() {
+          @Override
+          public Iterable<String> apply(Iterable<String> input) {
+            ArrayList<String> output = new ArrayList<>();
+            for (String element : input) {
+              output.add(element + "2");
+            }
+            return output;
+          }
+        };
+    batchingParDo = BatchingParDo.via(BATCH_SIZE, perBatchFn);
+  }
+
+  private static ArrayList<TimestampedValue<KV<String, String>>> createTestData() {
+    String[] scientists = {
+      "Einstein",
+      "Darwin",
+      "Copernicus",
+      "Pasteur",
+      "Curie",
+      "Faraday",
+      "Newton",
+      "Bohr",
+      "Galilei",
+      "Maxwell"
+    };
+    ArrayList<TimestampedValue<KV<String, String>>> data = new ArrayList<>();
+    long offset = 0;
+    for (int i = 0; i < NUM_ELEMENTS; i++) {
+      int index = i % scientists.length;
+      TimestampedValue<KV<String, String>> element =
+          TimestampedValue.of(
+              KV.of("key", scientists[index]),
+              startInstant.plus(Duration.standardSeconds(offset * TIMESTAMP_INTERVAL)));
+      data.add(element);
+      offset++;
+    }
+    return data;
+  }
+
+  // timer is not supported by DoFnTester
+  @Ignore
+  @Test
+  public void testUnderlyingDoFn() throws Exception {
+    DoFnTester<KV<String, String>, KV<String, String>> fnTester =
+        DoFnTester.of(
+            new BatchingParDo.BatchingDoFn<>(
+                BATCH_SIZE,
+                perBatchFn,
+                new Duration(ALLOWED_LATENESS),
+                StringUtf8Coder.of(),
+                StringUtf8Coder.of()));
+    int nbElementsProcessed = 0;
+    for (TimestampedValue<KV<String, String>> element : data) {
+      fnTester.processElement(element.getValue());
+      nbElementsProcessed++;
+      List<KV<String, String>> output = fnTester.takeOutputElements();
+      // end of batch
+      if ((nbElementsProcessed % BATCH_SIZE) == 0) {
+        assertEquals(
+            String.format("since last batch we should have processed %s elements", BATCH_SIZE),
+            BATCH_SIZE,
+            output.size());
+        assertTrue(
+            "All elements since last batch should have been processed",
+            checkAllElementsProcessing(output, Processing.PROCESSED));
+      } else {
+        // not end of batch
+        assertTrue(
+            "we should have processed no elements since last batch",
+            checkAllElementsProcessing(output, Processing.UNPROCESSED));
+      }
+    }
+  }
+
+  @Test
+  @Category({RunnableOnService.class, UsesTimersInParDo.class})
```

Comment:
I have seen this category and I asked myself the same question. 
But now that I look at the pom of spark runner (that does not support state yet), for the runnable-on-service-tests, I see UsesStatefulParDo category in the exclusions. 
I added the category in BatchingParDo